### PR TITLE
IGNITE-13448 Ducktests selftest: time is synchronized on nodes

### PR DESCRIPTION
**Pull request by andrey-kuznetsov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/ducktests/tests/ignitetest/tests/smoke_test.py [hunk 1]
```diff
@@ -87,3 +87,28 @@ def test_zk_start_stop(self):
...
         zookeeper.start()
         zookeeper.stop()
+
+    @cluster(num_nodes=4)
+    @ignite_versions(str(DEV_BRANCH))
+    def test_ignite_clock_sync(self, ignite_version):
+        """
+        Tests that clocks are synchronized between nodes.
+        """
+        num_nodes = 4
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/smoke_test.py:97]: Why do you check only 4 nodes. We should check that every node in a cluster is configured correctly.

File: modules/ducktests/tests/ignitetest/tests/smoke_test.py [hunk 2]
```diff
@@ -87,3 +87,28 @@ def test_zk_start_stop(self):
         zookeeper = ZookeeperService(self.test_context, num_nodes=3)
         zookeeper.start()
         zookeeper.stop()
+
+    @cluster(num_nodes=4)
+    @ignite_versions(str(DEV_BRANCH))
+    def test_ignite_clock_sync(self, ignite_version):
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/smoke_test.py:93]: It's not a smoke test of services. It checks that environment is configured correctly. So it's named as self tests. Actually there is a one test with the same goal - assertion_test. Let's rename it to smth like self_tests.py and move this test to the file.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/smoke_test/ClockSyncCheckerApplication.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+        IgniteCache<T2<UUID, Long>, Optional<Void>> cache = ignite.getOrCreateCache(
+            new CacheConfiguration<T2<UUID, Long>, Optional<Void>>(TIMESTAMP_CACHE)
+                .setCacheMode(CacheMode.REPLICATED));
+
+        cache.clear();
+
+        markInitialized();
+
+        ignite.cluster().forServers().nodes().forEach(server ->
```
Comment (timoninmaxim) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/smoke_test/ClockSyncCheckerApplication.java:78]: Looks like over complicated. Ducktape provide SSH account. So we can just use it for getting local time without start of ignite at all.

Actually is it possible to configure local time for an Ignite instance?

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

**Comment by timoninmaxim** [comment]
Also, python check style is broken.

File: modules/ducktests/tests/ignitetest/tests/smoke_test.py (hunk 2 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/smoke_test.py:93]: Fixed

File: modules/ducktests/tests/ignitetest/tests/smoke_test.py (hunk 1 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/smoke_test.py:97]: Fixed

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/smoke_test/ClockSyncCheckerApplication.java (hunk 3 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/smoke_test/ClockSyncCheckerApplication.java]: > Actually is it possible to configure local time for an Ignite instance?
Nope.

The rest is rewritten without the use of Ignite.

File: modules/ducktests/tests/ignitetest/tests/self_test.py [hunk 4]
```diff
@@ -53,3 +55,38 @@ def test_assertion_convertion(self, ignite_version):
...
+    """
+    Collects clock timestamps from a set of nodes.
+    """
+
+    def __init__(self, context, num_nodes):
+        super().__init__(context, num_nodes)
+        self.clocks = {}
+
+    def start_node(self, node):
+        self.clocks[node.name] = int(node.account.ssh_output("date +%s"))
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/self_test.py:89]: I'm not sure about name, is it unique among the cluster. So let's add additional check that len(res) equals to number of nodes.

File: modules/ducktests/tests/ignitetest/tests/smoke_test.py (hunk 2 above)
Comment (timoninmaxim) [reply modules/ducktests/tests/ignitetest/tests/smoke_test.py:93]: Let's remove this test from file.

File: modules/ducktests/tests/ignitetest/tests/self_test.py (hunk 4 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/self_test.py]: Agree, no uniqueness guaranteed. Turned dict into list.

File: modules/ducktests/tests/ignitetest/tests/self_test.py [hunk 5]
```diff
@@ -53,3 +55,38 @@ def test_assertion_convertion(self, ignite_version):
...
+    """
+    Collects clock timestamps from a set of nodes.
+    """
+
+    def __init__(self, context, num_nodes):
+        super().__init__(context, num_nodes)
+        self.clocks = []
+
+    def start_node(self, node):
+        self.clocks.append((node.name, int(node.account.ssh_output("date +%s"))))
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/self_test.py:89]: `date +%s` returns seconds. But max_clock_spread_ms is milliseconds.

Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/self_test.py]: Fixed.

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py [hunk 6]
```diff
@@ -47,6 +47,10 @@ def __init__(self, context, config, num_nodes, **kwargs):
 
         self.spec = resolve_spec(self, context, config, **kwargs)
 
+        # Force eager SSH client creation.
+        for node in self.nodes:
+            node.account.ssh("/bin/true")
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py:52]: Can be replaced with `node.account.ssh_client`, it is enough to client will be initialized

File: modules/ducktests/tests/ignitetest/tests/self_test.py [hunk 7]
```diff
@@ -53,3 +57,53 @@ def test_assertion_convertion(self, ignite_version):
...
+    Collects clock timestamps from a set of nodes.
+    """
+
+    def __init__(self, context, num_nodes):
+        super().__init__(context, num_nodes)
+        self.start_barrier = threading.Barrier(num_nodes)
+        self.clocks = []
+
+    def _worker(self, _, node):
+        node.account.ssh("/bin/true")
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/self_test.py:94]: Remove it now

File: modules/ducktests/tests/ignitetest/tests/self_test.py [hunk 8]
```diff
@@ -53,3 +57,53 @@ def test_assertion_convertion(self, ignite_version):
...
+
+        self.start_barrier.wait(1)
+        start = time.time()
+        delay = 5
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/self_test.py:98]: What is a purpose of this delay? We can just use `time.time() - start`, as delay is a constant. 

Actually to make a correction we should to run void command, like `ssh ;`.

File: modules/ducktests/tests/ignitetest/tests/self_test.py [hunk 9]
```diff
@@ -53,3 +57,53 @@ def test_assertion_convertion(self, ignite_version):
         else:
             app.stop()
             assert False
+
+    def test_clock_sync(self):
+        """
+        Tests if clocks are synchronized between nodes.
+        """
+
+        service = GetTimeService(self.test_context, self.test_context.cluster.num_available_nodes())
```
Comment (timoninmaxim) [inline modules/ducktests/tests/ignitetest/tests/self_test.py:66]: I think we should provide the check multiple times and compare max difference with `max_clock_spread_ms`.

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py (hunk 6 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py:52]: Done.

File: modules/ducktests/tests/ignitetest/tests/self_test.py (hunk 7 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/self_test.py:94]: It's not an IgniteAwareService.

File: modules/ducktests/tests/ignitetest/tests/self_test.py (hunk 8 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/self_test.py:98]: Thanks. It's obsolete now. Removing.

File: modules/ducktests/tests/ignitetest/tests/self_test.py (hunk 7 above)
Comment (timoninmaxim) [reply modules/ducktests/tests/ignitetest/tests/self_test.py:94]: Yep, this is true. OK

File: modules/ducktests/tests/ignitetest/tests/self_test.py (hunk 9 above)
Comment (andrey-kuznetsov) [reply modules/ducktests/tests/ignitetest/tests/self_test.py:66]: Done.

**Pull request closed without merging**